app/models/user.rb

```diff
@@ -4,10 +4,36 @@ class User < ApplicationRecord
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
 
-  has_one :homepage, dependent: :destroy
-  after_create -> { self.create_homepage }
+  # Virtual attribute for authenticating by either username or email
+  # This is in addition to a real persisted field like 'username'
+  attr_accessor :login
 
+  validates_format_of :username, with: /^[a-zA-Z0-9_\.]*$/, multiline: true
+  validates :username,
+            presence: true,
+            uniqueness: { case_sensitive: false }
+
+  has_one :homepage, dependent: :destroy
   has_many :lists
 
+  before_save { self.email.downcase! if self.email }
+  after_create -> { self.create_homepage }
+
   acts_as_voter
+
+  def self.find_for_database_authentication(warden_conditions)
+    conditions = warden_conditions.dup
+    if login = conditions.delete(:login)
+      where(conditions.to_h).where(["lower(username) = :value OR lower(email) = :value", { :value => login.downcase }]).first
+    elsif conditions.has_key?(:username) || conditions.has_key?(:email)
+      conditions[:email].downcase! if conditions[:email]
+      conditions[:username].downcase! if conditions[:username]
```

Comment:
So we should probably downcase the username before save, too, if we're gonna do this.